### PR TITLE
INT-3772: Add `MessagePreparedStatementCallback`

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-3772

### CODE REVIEW

File: spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcMessageHandler.java [hunk 1]
```diff
@@ -83,56 +109,106 @@ public JdbcMessageHandler(JdbcOperations jdbcOperations, String updateSql) {
...
+		}
+	}
+
 	/**
 	 * Executes the update, passing the message into the {@link SqlParameterSourceFactory}.
 	 */
 	@Override
-	protected void handleMessageInternal(Message<?> message) throws MessageRejectedException, MessageHandlingException,
-			MessageDeliveryException {
+	protected void handleMessageInternal(Message<?> message) throws Exception {
```
Comment (garyrussell) [inline spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcMessageHandler.java]: Why the change to a checked exception?

File: spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcOutboundGateway.java [hunk 2]
```diff
@@ -181,6 +183,11 @@ public void setRequestSqlParameterSourceFactory(SqlParameterSourceFactory sqlPar
 		this.handler.setSqlParameterSourceFactory(sqlParameterSourceFactory);
 	}
 
+	public void setRequestPreparedStatementSetter(MessagePreparedStatementSetter requestPreparedStatementSetter) {
+		Assert.notNull(this.handler);
```
Comment (garyrussell) [inline spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcOutboundGateway.java]: 'handler' cannot be null

File: spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcMessageHandler.java [hunk 3]
```diff
@@ -83,56 +109,106 @@ public JdbcMessageHandler(JdbcOperations jdbcOperations, String updateSql) {
...
 	 * @param keysGenerated the flag value to set
 	 */
 	public void setKeysGenerated(boolean keysGenerated) {
 		this.keysGenerated = keysGenerated;
 	}
 
-	public void setUpdateSql(String updateSql) {
-		this.updateSql = updateSql;
-	}
-
```
Comment (garyrussell) [inline spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcMessageHandler.java]: I don't think we should take away the ability to change the sql at runtime.

**Comment by garyrussell** [comment]
Polish here: https://github.com/garyrussell/spring-integration/commit/ea90becda8ad449f45a560ebe907d29361dd4b40

Just need a resolution to the updateSql issue above.

**Comment by artembilan** [comment]
> I don't think we should take away the ability to change the sql at runtime.

Well it is some polemic place. 
Our current practice is: do not allow change properties which are populated from the ctor. And finally they are `final` anyway. Right, I made that change here and now it isn't possible. I'm not sure that my optimization with `generatedKeysStatementCreator` instance will work with the non-`final` property.

We can come back, of course, to the old style and `volatile` `updateSql`, or just can make a `Migration Guide` note that it isn't possible right now by the performance reason.

WDYT?

**Comment by garyrussell** [comment]
I just don't like removing features that people might (albeit unlikely) be using.

Can't we simply recreate the `generatedKeysStatementCreator` if the updateSql changes?

**Comment by artembilan** [comment]
> Can't we simply recreate the generatedKeysStatementCreator if the updateSql changes?

Yes, sounds good. Let me fix that after `scanner` PR.
With applying your polishing, of course :smile:

**Comment by artembilan** [comment]
Pushed.

**Comment by garyrussell** [comment]
Merged as 9ff5ab789a9df9dbe127eb917d5f5cb739e1b585 after rebase
